homeassistant/helpers/template.py

```diff
@@ -246,11 +271,13 @@ def __init__(self, hass, domain):
 
     def __getattr__(self, name):
         """Return the states."""
-        return _wrap_state(
-            self._hass.states.get('{}.{}'.format(self._domain, name)))
+        return _get_state(
+            self._hass, '{}.{}'.format(self._domain, name))
 
     def __iter__(self):
         """Return the iteration over all the states."""
+        if _ENTITY_COLLECT is not None:
```

Comment:
I think that this approach is overkill. 

I think that if people are doing super dynamic things, they should just specify their entity IDs. It's impossible to figure out what entities are going to be impacted. 

We also should not use a global but instead, since every template is being compiled once and generates all these values, we should just store it in the template.

Reply:
It's more a thread local really, only we know empirically that it's just a single thread going through this code, and it's single threaded as it's on the event loop. 

It is possible to figure out what entities are impacted in a general sense, which is exactly that this PR does. It keeps a list of all the entities that were hit in the run through, and if their state doesn't change then the overall template value can't change either. When the states change then you check the template again to get the result, and again keep a list of anything impacted for the next trigger.

For example a template like `{{ state1 or state2 }}` can't possibly change state unless state1 is true, regardless of what state2 does. If state1 becomes true, the recheck is triggered, and then state1 and state2 are on the list for next time.

Reply:
Oh, just saw where you were talking about. If the template uses the iterators, then it touches every state (or every domain state) to calculate the result. If any of those change then it will potentially change the output. I wanted to get away from specifying entity IDs manually since that's going to be a rarely used and thus poorly understood and used feature.

It's only an issue if you iterate in the template, but in the end we're accumulating a filter we might as well go the whole hog and do it properly here, there's not a lot of good reason not to since it's fairly straightforward to do this.

Reply:
So the change here is that we don't try to figure out what is _going_ to be impacted. We record the fact as entities _are_ impacted. That's very intriguing. Listing all entities can be a real chore and error prone when you are, for example, iterating domains. It's also something that must be constantly maintained as your setup grows.

On the other hand, I think this could be inefficient. Say a template like this,
```
{% for state in states.sensor %}
  {% if state.entity_id | regex_match('.*humidity') %}
    {{ state.entity_id }}={{ state.state }}
  {% endif %}
{% endfor %}
```
This would have to watch all sensors even though a human might know that only the humidity ones can affect the result. Also, when adding a new entity we must assume that it will affect all templates.

Those possible issues are not meant to shoot down this proposal. Like I said, I am very intrigued and I think it deserves serious consideration. Right now I do have template sensors that don't work properly because I did not bother listing all affected entities.

(I didn't look at the code changes)

Reply:
Yes, basically exactly.

In the example you gave you'd filter on the sensor domain, so it's not a disaster. Besides in that template you've really got no choice *but* to wait for changes on everything, because we can never know a priori all of the entities it might hit. Even at template registration time we don't actually know, because that will be during the hass initialisation sequence and potentially more states will be added after registering the template.

This will do what you want and just wait on those states:

```
{% for st in ['bathroom_humidity', 'outside_humidity'] %}
  {{ state.sensor(st).entity_id }}={{ state.sensor(st).state }}
{% endfor %}
```

This fairly egregariously daft template will listen for any state change, but it's pretty daft:

```
{% for state in states %}
  {% if state.entity_id == 'state.humidity' %}
    {{ state.entity_id }}={{ state.state }}
  {% endif %}
{% endfor %}
```

Anyhow, even at the very worst you have an evaluation of the template with each state change, it's not *desirable* but compared to the work going on with say image processing which happens multiple times a second it's not huge, and you'd really have to write a deliberately daft template to even cause that to happen.

To go back to the old way of specifying the list of entity IDs you're going to cause a lot of issues that are going to be subtle and hard to debug.

Reply:
Eh, it's not because I am really slow ... I somehow missed your comment above mine where you make all the same points :-)

Reply:
Probably warrants re-explanation anyhow. I felt pretty clever when I came up with the idea 😛